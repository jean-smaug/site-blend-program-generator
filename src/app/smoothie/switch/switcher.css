.switcher {
  position: fixed;
  top: 100px;
  left: 0;
  background: grey;
  width: 200px;
}

.switcher__item {
  background: white;
  cursor: pointer;
}
