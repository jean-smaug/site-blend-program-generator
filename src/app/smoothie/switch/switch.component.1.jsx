// @flow

import React, { Component } from 'react';
import _ from 'lodash';
import { connect } from 'react-redux';
import { switchConference } from '../smoothie.action';

import { Conference, Conferences } from '../smoothie.type';
import './switcher.css';

export class SwitchComponent extends Component {
  props: {
    switchConference: (conference: Conference) => void,
    conferences: Conferences,
    currentConferenceId: number,
  };

  selectConference = (conference: Conference) => {
    this.props.switchConference(conference);
  };

  render() {
    const { conferences, currentConferenceId } = this.props;

    const remainingConferences = _.filter(conferences, item => item.id !== currentConferenceId);
    return (
      <div>
        <div className="switcher">
          <ul>
            {_.map(remainingConferences, item => (
              <li
                className="switcher__item"
                key={item.id}
                role="presentation"
                onClick={() => this.selectConference(item)}
              >
                {item.title}
              </li>
            ))}
          </ul>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  currentConferenceId: state.smoothie.currentConferenceId,
});

const mapDispatchToProps = dispatch => ({
  switchConference: conference => dispatch(switchConference(conference)),
});

export default connect(mapStateToProps, mapDispatchToProps)(SwitchComponent);
