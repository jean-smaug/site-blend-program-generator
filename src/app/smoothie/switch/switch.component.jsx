// @flow
/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */

import { connect } from 'react-redux';
import React, { Component } from 'react';
import _ from 'lodash';
import { Conference, Conferences } from '../smoothie.type';
import { switchConference } from '../smoothie.action';
import './switcher.css';
import { getEndTime, convertHourToString } from '../../../lib/time.lib';
import { getConferencesConflict } from '../../../lib/dataFilter.lib';

// import { Conferences } from '../../smoothie/smoothie.type';

export class SwitcherComponent extends Component {
  state = {
    conferenceSelected: null,
    conferencesConflits: null,
  };

  state: {
    conferenceSelected: Conference,
    conferencesConflits: Conferences,
  };

  props: {
    conferences: Conferences,
    closeModal: () => void,
    smoothie: Conferences,
    switchConference: (conference: Conference) => void,
  };

  switchConference = (conference: Conference) => {
    const time = conference.timeBegin.split('h')[0];
    const letterTime = convertHourToString(time);
    const timeSlotConferences = this.props.smoothie[conference.day][letterTime];
    const conferencesConflits = getConferencesConflict(
      timeSlotConferences.selected,
      conference,
    ).map(item => <li>{item.title}</li>);
    this.setState({
      conferencesConflits,
      conferenceSelected: conference,
    });
  };

  confirmSelection = async() => {
    this.props.switchConference(this.state.conferenceSelected);
    this.props.closeModal();
  };

  render() {
    return (
      <div className="modal is-active">
        <div
          className="modal-background"
          onClick={this.props.closeModal}
          role="button"
          aria-pressed="true"
          tabIndex="0"
        />
        <div className="modal-card">
          <section className="modal-card-header header-restore">
            <h1>Switcher de conférence !</h1>
          </section>
          <section className="modal-card-body">
            {this.state.conferencesConflits !== null ? (
              <div className="message-body">
                <p className="msg-conflit">Attention les conférences suivantes sont en conflit horaire avec la séléction et
                vont donc être enlevées du planning :</p>
                {this.state.conferencesConflits}
              </div>
            ) : null}
            <ul>
              {_.map(this.props.conferences, (conference, id) => (
                <li
                  className={
                    this.state.conferenceSelected === conference ? (
                      'item-switcher selected-conf'
                    ) : (
                      'item-switcher'
                    )
                  }
                  role="presentation"
                  key={id}
                  onClick={() => this.switchConference(conference)}
                >
                  <p>{`${conference.timeBegin} > ${getEndTime(
                    conference.timeBegin,
                    conference.duration,
                  )}`}</p>
                  <p>{conference.title.charAt(0).toUpperCase() +
                  conference.title.substring(1).toLowerCase()}</p>
                </li>
              ))}
            </ul>
            <input
              type="button"
              className="btn-suivant"
              onClick={this.confirmSelection}
              value="confirmer"
            />
          </section>
          <button className="modal-close is-large" onClick={this.props.closeModal} />
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  smoothie: state.smoothie,
});

const mapDispatchToProps = dispatch => ({
  switchConference: conference => dispatch(switchConference(conference)),
});

export default connect(mapStateToProps, mapDispatchToProps)(SwitcherComponent);
