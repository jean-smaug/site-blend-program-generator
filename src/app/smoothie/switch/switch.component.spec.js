import React from 'react';
import { shallow } from 'enzyme';

import { SwitchComponent } from './switch.component';

const conferences = [{}];

describe('switch component', () => {
  it('should render the switch component', () => {
    const Switch = shallow(<SwitchComponent currentConferenceId={0} conferences={conferences} />);
    expect(Switch).toMatchSnapshot();
  });

  // it('detect click button', () => {
  //   const Switch = shallow(
  //     <SwitchComponent id={0} conferences={conferences} />,
  //   );
  // });
});
