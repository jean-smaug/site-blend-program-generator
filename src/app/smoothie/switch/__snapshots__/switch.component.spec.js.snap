// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`switch component should render the switch component 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <div
        className="switcher"
    >
        <ul>
            <li>
                <button
                    onClick={[Function]}
                />
            </li>
        </ul>
    </div>
</div>,
  "nodes": Array [
    <div>
      <div
            className="switcher"
      >
            <ul>
                  <li>
                        <button
                              onClick={[Function]}
                        />
                  </li>
            </ul>
      </div>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <SwitchComponent
        conferences={
                Array [
                        Object {},
                      ]
        }
        currentConferenceId={0}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": SwitchComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "conferences": Array [
            Object {},
          ],
          "currentConferenceId": 0,
        },
        "refs": Object {},
        "selectConference": [Function],
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          <div
                    className="switcher"
          >
                    <ul>
                              <li>
                                        <button
                                                  onClick={[Function]}
                                        />
                              </li>
                    </ul>
          </div>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          <div
                    className="switcher"
          >
                    <ul>
                              <li>
                                        <button
                                                  onClick={[Function]}
                                        />
                              </li>
                    </ul>
          </div>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <SwitchComponent
    conferences={
        Array [
            Object {},
          ]
    }
    currentConferenceId={0}
/>,
}
`;
